v1/src/main/java/com/google/cloud/teleport/spanner/AvroRecordConverter.java

```diff
@@ -374,6 +384,45 @@ static Optional<List<String>> readStringArray(
     }
   }
 
+  @VisibleForTesting
+  @SuppressWarnings("unchecked")
+  static Optional<List<Float>> readFloat32Array(
+      GenericRecord record, Schema.Type avroType, String fieldName) {
+    Object fieldValue = record.get(fieldName);
+    if (fieldValue == null) {
+      return Optional.empty();
+    }
+    switch (avroType) {
+        // For type check at compile time, the type of x has to be specified (as cast) so that
+        // convertability to float can be verified.
+      case FLOAT:
+        return Optional.of((List<Float>) fieldValue);
+      case INT:
+        // Can cause loss of precision: un-representable integers will be rounded to the nearest
+        // float.
+        {
+          List<Integer> value = (List<Integer>) fieldValue;
+          return Optional.of(
+              value.stream()
+                  .map(x -> x == null ? null : Float.valueOf(x))
+                  .collect(Collectors.toList()));
+        }
+      case STRING:
+        {
+          List<Utf8> value = (List<Utf8>) record.get(fieldName);
+          return Optional.of(
+              value.stream()
+                  .map(x -> x == null ? null : Float.valueOf(x.toString()))
```

Comment:
What happens when the cast is not possible?

Reply:
It throws an error, and the behavior would be similar to what happens in FLOAT64. The parsing will fail when the string is not a numeric.